### ISSUE TITLE
Extend SpineOpt with different algorithms

### ISSUE THREAD

**Issue by manuelma** [spine-tools/SpineOpt.jl#820]
Right now if I want to try a new way of solving a SpineOpt model, I need to modify SpineOpt source code. I would like to be able to do it from my client script, because my new solution method may not be intended to be part of SpineOpt.

What I propose is to introduce a new keyword argument to run_spineopt so I can specify a function to call when it's time to solve the model. The default would be `run_spineopt_kernel!` which is already part of SpineOpt.

Similarly, if I need new 'template' data to support my algorithm, at the moment my only option is to modify the SpineOpt official template. I would like to be able to provide additional custom templates via run_spineopt too.

**Comment by tarskul** [comment]
Will this functionality then also cover the existing algorithms (i.e. Benders and MGA)? Will the model type parameter in the spine db be used to choose between the existing algorithms and your own algorithms?

For the template I was also thinking about ways to provide different templates (and adjust the load_template tool in spine toolbox accordingly). One of the things I'm considering is whether users would be able to use this feature to combine the default structure (classes, parameter definitions, ...) a model structure (algorithm, temporal and stochastic structure) and an energy system (e.g. a cogeneration unit). But issues may arise if these templates are out of date or out of sync so they may need some sort of versioning (or at the very least, all the templates we provide are part of the unit tests and any additional file the user finds or makes is their own responsibility).

**Comment by manuelma** [comment]
> Will this functionality then also cover the existing algorithms (i.e. Benders and MGA)? Will the model type parameter in the spine db be used to choose between the existing algorithms and your own algorithms?

No, the existing algorithms will continue to be part of SpineOpt. I am talking only of extending/replacing via code the lowest-level algorithm (basically, the one that rolls over the temporal structure while optimising the model at every window, carrying results, etc). I call this the *kernel* and my idea is to make all the current algorithms (standard, benders, and mga) to call this kernel whenever they need to solve the model. And top of that, allow users to override the kernel via a function so they try other things. I hope that's clear, thanks for the question.
